py/parsenum.c

```diff
@@ -295,7 +295,17 @@ mp_obj_t mp_parse_num_decimal(const char *str, size_t len, bool allow_imag, bool
             exp_val -= SMALL_NORMAL_EXP;
             dec_val *= SMALL_NORMAL_VAL;
         }
-        dec_val *= MICROPY_FLOAT_C_FUN(pow)(10, exp_val);
+
+        // At this point, we need to multiply the mantissa by its base 10 exponent. If possible,
+        // we would rather manipulate numbers that have an exact representation in IEEE754. It
+        // turns out small positive powers of 10 do, whereas small negative powers of 10 don't.
+        // So in that case, we'll yield a division of exact values rather than a multiplication
+        // of slightly erroneous values.
+        if (exp_val < 0 && exp_val > -15) { // 15 = log10(2^53)
```

Comment:
Remember that this code also needs to work with single precision floats, so -15 would be something like -6 if single precision is used.

Reply:
Indeed. Would you consider the merge if I fixed this?

Reply:
Great, makes perfect sense. If you decide you want to merge it, i'll be happy to make it work with single-precision floats too (well, it's really just an #if statement…) 😄